composer/trainer/trainer.py

```diff
@@ -582,20 +583,21 @@ def __init__(
             steps_per_epoch=train_subset_num_batches,
         )
 
-        schedulers = ensure_tuple(schedulers)
-        if len(schedulers) == 0:
-            schedulers = (constant_scheduler,)
-            warnings.warn(f"No scheduler was specified. Defaulting to {repr(schedulers)}")
-
-        if scale_schedule_ratio != 1.0:
-            schedulers = tuple(
-                scale_scheduler(scheduler, scale_schedule_ratio) for scheduler in ensure_tuple(schedulers))
-
         if step_schedulers_every_batch is None:
-            step_schedulers_every_batch = any(callable(scheduler) for scheduler in schedulers)
+            # only if all schedulers are ComposerSchedulers, then we can step every batch by default
+            step_schedulers_every_batch = all(
+                not isinstance(scheduler, PytorchScheduler) for scheduler in ensure_tuple(schedulers))
```

Comment:
This is another place where I'm conflicted, here on `any` vs `all`. On the one hand I'd argue that someone use at least one of our `ComposerScheduler` objects should not be surprised by our batchwise stepping. On the other hand, I could see someone being annoyed by having to reconfigure their `PytorchScheduler` objects when they add or remove a `ComposerScheduler`.

Reply:
Yeah, I suspect that having one scheduler's behavior change depending on the behavior of others will cause someone to get really confused at some point. E.g., "why are my results different when I add in a composer ConstantLR with lr=1.0, which in theory should do nothing?".

Can we just step each scheduler based on its own preference? Might need a LambdaLR subclass with a `def step_every_batch(self) -> bool: return True` method or something (that `compile_composer_scheduler` would use).

Reply:
Simple solution would be to just fail fast if different schedulers have different preferences.

Reply:
I am leaning towards where it would be more incorrect that a epochwise scheduler is stepped every batch (compared to the loss of precision by stepping every epoch), so I am inclined to require that all schedulers be composer schedulers if this is not set. Added detailed log messages here to explain the behavior.